feat: introduce new deprecated types for rules (#19238)
eslint/eslint: docs/src/extend/rule-deprecation.md

---
The rule deprecation metadata describes whether a rule is deprecated and how the rule can be replaced if there is a replacement.
The legacy format used the two top-level [rule meta](./custom-rules.md#rule-structure) properties `deprecated: true` and `replacedBy`.
In the new format `deprecated` is an object of type `DeprecatedInfo` and `replacedBy` should be defined inside `deprecated` instead of the top-level.

## ◆ DeprecatedInfo type
This type represents general information about a rule deprecation.
Every property is optional.

* `message` (`string`)<br>
  A general message presentable to the user. May contain why this rule is deprecated or how to replace the rule.
* `url` (`string`)<br>
  An URL with more information about this rule deprecation.
* `replacedBy` (`ReplacedByInfo[]`)<br>
  Information about the available replacements for the rule.
  This may be an empty array to explicitly state there is no replacement.
* `deprecatedSince` (`string`)<br>
  [Semver](https://semver.org/) of the version deprecating the rule.
* `availableUntil` (`string | null`)<br>
  [Semver](https://semver.org/) of the version likely to remove the rule, e.g. the next major version.

>   The special value `null` means the rule is "frozen". This means the rule will no longer be changed but will be kept available.

Reviewer: [The RFC](https://github.com/eslint/rfcs/blob/main/designs/2024-deprecated-rule-metadata/README.md?plain=1#L62) does indeed state that a value of `null` for `deprecated.availableUntil` will indicate 
a frozen rule, but the new implementation doesn't account for that use case. If we want to account for frozen rules in the implementation then I think we should:

1. Add a `meta.deprecated` property to frozen rules, with `meta.deprecated.availableUntil` set to `null`.
2. Update the way we check for deprecated rules (i.e. rules that are deprecated but not frozen). Instead of testing just the truthiness of `meta.deprecated`, we should also test that `meta.deprecated.availableUntil` isn't `null`. There are a few places where this kind of check is performed, both in production code and in the build, and each check would have to be updated.

I'm not sure if this is what we want, so I wouldn't be opposed to leaving out the mention of the special `null` value from the rule deprecation docs and ignoring the case of frozen rules for now. Thoughts? @eslint/eslint-tsc

Reviewer 2: My understanding of "frozen" (`availableUntil:null`) in the RFC was that the rule is deprecated but won't be removed. We're using the term "frozen" for rules that are not deprecated but are not accepting new feature requests, so I don't think we should add `meta.deprecated` to those rules since they are not deprecated.

I think it would be best to remove the word "frozen" from this description to avoid confusion:

```suggestion
  The special value `null` means the rule will no longer be changed but will be kept available.
```

* `availableUntil` (`string | null`)<br>
  [Semver](https://semver.org/) of the version likely to remove the rule, e.g. the next major version.
  The special value `null` means the rule will no longer be changed but will be kept available.

## ◆ ReplacedByInfo type

The type describes a single possible replacement of a rule.
Every property is optional.

* `message` (`string`)<br>
  A general message about this rule replacement, e.g.
* `url` (`string`)<br>
  An URL with more information about this rule replacement.
* `plugin` (`ExternalSpecifier`)<br>
  Specifies which plugin has the replacement rule.
  The name should be the package name and should be "eslint" if the replacement is an ESLint core rule.
  This property should be omitted if the replacement is in the same plugin.
* `rule` (`ExternalSpecifier`)<br>
  Specifies the replacement rule.
  May be omitted if the plugin only contains a single rule or has the same name as the rule.

### ◆ ExternalSpecifier type

This type represents an external resource.
Every property is optional.

* `name` (`string`)<br>
  The package name for `plugin` and the rule id for `rule`.
* `url` (`string`)<br>

>   An URL pointing to additional information, for example an announcement post or migration guide.

Reviewer 2: ```suggestion
  An URL pointing to documentation for the plugin / rule..
```
